File: 864. Shortest Path to Get All Keys/Solution.java
class Solution {
    public int shortestPathAllKeys(String[] grid) {
        int m = grid.length;
        int n = grid[0].length();
        int allKeys = 0;
        int startRow = -1;
        int startCol = -1;
        int[][] directions = {{-1, 0}, {1, 0}, {0, -1}, {0, 1}};
        for (int i = 0; i < m; i++) {
            for (int j = 0; j < n; j++) {
                char c = grid[i].charAt(j);
                if (c == '@') {
                    startRow = i;
                    startCol = j;
                } else if (c >= 'a' && c <= 'f') {
                    allKeys |= (1 << (c - 'a'));
                }
            }
        }
        Queue<int[]> queue = new LinkedList<>();
        boolean[][][] visited = new boolean[m][n][64];
        queue.offer(new int[]{startRow, startCol, 0});
        visited[startRow][startCol][0] = true;
        int steps = 0;
        while (!queue.isEmpty()) {
            int size = queue.size();
            while (size-- > 0) {
                int[] curr = queue.poll();
                int row = curr[0];
                int col = curr[1];
                int keys = curr[2];
                if (keys == allKeys) {
                    return steps;
                }
                for (int[] dir : directions) {
                    int newRow = row + dir[0];
                    int newCol = col + dir[1];
                    if (newRow >= 0 && newRow < m && newCol >= 0 && newCol < n) {
                        char cell = grid[newRow].charAt(newCol);
                        if (cell == '#') {
                            continue;
                        }
                        int newKeys = keys;
                        if (cell >= 'a' && cell <= 'f') {
                            newKeys |= (1 << (cell - 'a'));
                        }
                        if (cell >= 'A' && cell <= 'F' && (keys & (1 << (cell - 'A'))) == 0) {
                            continue;
                        }
                        if (!visited[newRow][newCol][newKeys]) {
                            queue.offer(new int[]{newRow, newCol, newKeys});
                            visited[newRow][newCol][newKeys] = true;
                        }
                    }
                }
            }
            steps++;
        }
        return -1;
    }
}

File: 864. Shortest Path to Get All Keys/Solution.py
class Solution:
    def shortestPathAllKeys(self, grid):
        m = len(grid)
        n = len(grid[0])
        allKeys = 0
        startRow = -1
        startCol = -1
        directions = [[-1, 0], [1, 0], [0, -1], [0, 1]]
        for i in range(m):
            for j in range(n):
                c = grid[i][j]
                if c == '@':
                    startRow = i
                    startCol = j
                elif c >= 'a' and c <= 'f':
                    allKeys |= (1 << (ord(c) - ord('a')))
        queue = deque()
        visited = [[[False] * 64 for _ in range(n)] for _ in range(m)]
        queue.append([startRow, startCol, 0])
        visited[startRow][startCol][0] = True
        steps = 0
        while queue:
            size = len(queue)
            while size > 0:
                curr = queue.popleft()
                row = curr[0]
                col = curr[1]
                keys = curr[2]
                if keys == allKeys:
                    return steps
                for dir in directions:
                    newRow = row + dir[0]
                    newCol = col + dir[1]
                    if newRow >= 0 and newRow < m and newCol >= 0 and newCol < n:
                        cell = grid[newRow][newCol]
                        if cell == '#':
                            continue
                        newKeys = keys
                        if cell >= 'a' and cell <= 'f':
                            newKeys |= (1 << (ord(cell) - ord('a')))
                        if cell >= 'A' and cell <= 'F' and (keys & (1 << (ord(cell) - ord('A')))) == 0:
                            continue
                        if not visited[newRow][newCol][newKeys]:
                            queue.append([newRow, newCol, newKeys])
                            visited[newRow][newCol][newKeys] = True
                size -= 1
            steps += 1
        return -1